### PR TITLE
CBG-2929 add HTTPBytesWritten stat for public API

### PR DESCRIPTION
**Pull request by torcolvin**
CBG-2929

Created a custom http.ResponseWriter and log the DB scoped requests over the public and guest APIs.

## Pre-review checklist
- [x] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [x] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [x] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/1928/

### CODE REVIEW

File: base/stats.go [hunk 1]
```diff
@@ -461,6 +461,8 @@ type DatabaseStats struct {
 	DocWritesXattrBytes *SgwIntStat `json:"doc_writes_xattr_bytes"`
 	// Highest sequence number seen on the caching DCP feed.
 	HighSeqFeed *SgwIntStat `json:"high_seq_feed"`
+	// Number of bytes written over public interface for REST api
```
Comment (adamcfraser) [inline base/stats.go]: This looks like it's just a subset of CBG-2929 (doesn't include request bytes over REST, and doesn't include blip).  If it makes sense to split into separate PRs, creating subtickets under CBG-2929 seems like a good idea to make sure nothing gets missed.

File: rest/counted_response_writer.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+// CountableResponseWriter is an interface that any custom http.ResponseWriter used by Sync Gateway handlers need to implement.
+type CountableResponseWriter interface {
+	http.ResponseWriter
+	GetBytesWritten() int64
+}
+
+// http.ResponseWriter wrapper that counts the number of bytes written in a response. This ignores the bytes written in headers.
+type CountedResponseWriter struct {
+	writer       http.ResponseWriter
+	bytesWritten int64
```
Comment (adamcfraser) [inline rest/counted_response_writer.go]: Would it be more efficient and/or generate less GC to pass the stat into CountedResponseWriter, and just update the stat directly instead of maintaining a separate int64 on the CountedResponseWriter?

File: rest/handler.go [hunk 3]
```diff
@@ -557,6 +559,20 @@ func (h *handler) logRequestLine() {
...
 
+// logBytesWritten only logs database level bytes written over the public API
+func (h *handler) logBytesWritten() {
+	if h.db == nil {
+		return
+	}
+	if h.privs != publicPrivs && h.privs != regularPrivs {
+		return
+	}
+	if h.isBlipSync() {
```
Comment (adamcfraser) [inline rest/handler.go:603]: How are we handling bytes associated with streaming non-blip _changes requests (0feed=continuous or feed=websocket)?

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: base/stats.go (hunk 1 above)
Comment (torcolvin) [reply base/stats.go]: These are in https://github.com/couchbase/sync_gateway/commit/a31be806b47b07cb9e3d7fdfaaaa8ec024d2f0a8

File: rest/handler.go (hunk 3 above)
Comment (torcolvin) [reply rest/handler.go:603]: Should be yes if they are over public API.

File: rest/counted_response_writer.go (hunk 2 above)
Comment (torcolvin) [reply rest/counted_response_writer.go]: I sort of did this and not, the problem is we have to create a the response writer before we know if the database exists or not, so we may or may not be doing anything with this value.

Comment (adamcfraser) [reply rest/counted_response_writer.go]: If this implies that we're associating a CountedResponseWriter with a request even if it's not a db-scoped request, that doesn't seem ideal.  What's the reason it needs to be initialized that early, as opposed to doing it once we've set the db on the handler?

File: rest/server_context.go [hunk 4]
```diff
@@ -41,8 +41,8 @@ const kStatsReportInterval = time.Hour
 const kDefaultSlowQueryWarningThreshold = 500 // ms
 const KDefaultNumShards = 16
 
-// defaultBlipStatsReportingInterval is the default interval when to report blip stats, at the end of a message handler.
-const defaultBlipStatsReportingInterval = 30 * time.Second
+// defaultBytesStatsReportingInterval is the default interval when to report blip stats, at the end of a message handler.
```
Comment (adamcfraser) [inline rest/server_context.go]:
```suggestion
// defaultBytesStatsReportingInterval is the default interval when to report request bytes transferred stats
```

File: rest/handler.go [hunk 5]
```diff
@@ -294,11 +303,11 @@ func (h *handler) validateAndWriteHeaders(method handlerMethod, accessPermission
 		return base.HTTPErrorf(http.StatusUnsupportedMediaType, "Unsupported Content-Encoding; use gzip")
 	}
 
-	if h.shouldShowProductVersion() {
-		h.setHeader("Server", base.VersionString)
-	} else {
-		h.setHeader("Server", base.ProductNameString)
-	}
+	//if h.shouldShowProductVersion() {
```
Comment (adamcfraser) [inline rest/handler.go]: I'm not sure what's going on with this change - is this intentional?

File: rest/handler.go [hunk 6]
```diff
@@ -582,6 +592,30 @@ func (h *handler) logRequestLine() {
 	base.InfofCtx(h.ctx(), base.KeyHTTP, "%s %s%s%s", h.rq.Method, base.SanitizeRequestURL(h.rq, &queryValues), proto, h.formattedEffectiveUserName())
 }
 
+// iupdateResponseWriter will create an updated Response Writer if we need to log db stats.
+func (h *handler) isPublicDBRequestForStats() bool {
```
Comment (adamcfraser) [inline rest/handler.go]: "isPublicDBRequestForStats" reads like "is this a request for stats over the public API".  

Is a better name for this just "shouldUpdateBytesTransferredStats"?  In the comment you could describe the criteria we're using for this, which I think is just "is a non-blip, database scoped call over the public port"

File: rest/api_test.go [hunk 7]
```diff
@@ -2275,6 +2276,7 @@ func TestHandleStats(t *testing.T) {
 
 // TestHideProductInfo ensures that detailed product info is not shown on non-admin REST API responses if set.
 func TestHideProductInfo(t *testing.T) {
+	t.Skip("here")
```
Comment (adamcfraser) [inline rest/api_test.go]: Unclear what's changed that's requiring this to be skipped.

File: rest/changes_api.go [hunk 8]
```diff
@@ -402,14 +402,6 @@ func (h *handler) sendSimpleChanges(channels base.Set, options db.ChangesOptions
 			}
 		}
```
Comment (adamcfraser) [inline rest/changes_api.go]: This change seems unrelated to this enhancement, if this kind of cleanup is justified I think it should be as a standalone PR.

File: rest/doc_api_test.go [hunk 9]
```diff
@@ -164,7 +164,7 @@ func TestGuestReadOnly(t *testing.T) {
 	// Attempt to access _blipsync as guest - blip sync handling for read-only GUEST is applied at replication level (to allow pull-only replications).
 	// Should succeed permission check, and only fail on websocket upgrade
 	response = rt.SendRequest("GET", "/{{.db}}/_blipsync", "")
-	RequireStatus(t, response, http.StatusUpgradeRequired)
+	RequireStatus(t, response, http.StatusInternalServerError)
```
Comment (adamcfraser) [inline rest/doc_api_test.go]: What's happening in this PR to change the error response for this endpoint?

File: rest/encoded_response_writer.go [hunk 10]
```diff
@@ -66,17 +78,37 @@ func (w *EncodedResponseWriter) WriteHeader(status int) {
...
+		w.reportStats(false)
+		return n, err
+	}
+	n, err := w.ResponseWriter.Write(b)
+	w.lastBytesWritten += int64(n)
+	w.reportStats(false)
+	return n, err
+}
+
+// ReportStats reports bytes written GetLastBytesWritten returns the number of bytes written by this response writer. This is not locked, so is only safe to call while no one is calling CountedResponseWriter.Write, usually after the response has been fully written.
```
Comment (adamcfraser) [inline rest/encoded_response_writer.go]:
```suggestion
// ReportStats updates the stat with the number of bytes written by this response writer. This is not locked, so is only safe to call while no one is calling CountedResponseWriter.Write, usually after the response has been fully written.
```

File: rest/encoded_response_writer.go [hunk 11]
```diff
@@ -66,17 +78,37 @@ func (w *EncodedResponseWriter) WriteHeader(status int) {
...
+	w.lastBytesWritten = 0
+	w.lastReportTime = currentTime
 }
 
 func (w *EncodedResponseWriter) disableCompression() {
 	w.sniffDone = true
 }
 
 func (w *EncodedResponseWriter) sniff(bytes []byte) {
+	fmt.Println(w)
```
Comment (adamcfraser) [inline rest/encoded_response_writer.go]: Why does this need to be added to sniff?

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: rest/changes_api.go (hunk 8 above)
Comment (torcolvin) [reply rest/changes_api.go]: It's not unrelated exactly, because I didn't implement deprecated (from go 1.7) CloseNotifier interface in the ResponseWriters.

I can pull it out into a separate commit though.

File: rest/encoded_response_writer.go (hunk 11 above)
Comment (torcolvin) [reply rest/encoded_response_writer.go]: mistake.

File: rest/doc_api_test.go (hunk 9 above)
Comment (torcolvin) [reply rest/doc_api_test.go]: Fixed with a new isHijackable method.

File: rest/api_test.go (hunk 7 above)
Comment (torcolvin) [reply rest/api_test.go]: mistake

File: rest/handler.go (hunk 5 above)
Comment (torcolvin) [reply rest/handler.go]: mistake

File: base/stats.go [hunk 12]
```diff
@@ -461,6 +461,8 @@ type DatabaseStats struct {
...
+	HTTPBytesWritten *SgwIntStat `json:"http_bytes_written"`
```
Comment (bbrks) [inline base/stats.go]: This seems ambiguous if we're not including Admin or Metric API data in it. Should this be specifically named after the public API?

Comment (bbrks) [inline base/stats.go]: Where are we accounting for request body sizes? (here seems like response body size only)

File: rest/encoded_response_writer.go [hunk 13]
```diff
@@ -121,6 +151,11 @@ func (w *EncodedResponseWriter) Close() {
 	}
 }
 
+// isHijackable determines if the underlying writer implements hijack interface
+func (w *EncodedResponseWriter) isHijackable() bool {
+	return false
```
Comment (bbrks) [inline rest/encoded_response_writer.go:158]: What makes this non-hijackable? Isn't there a potentially hijackable ResponseWriter inside the EncodedResponseWriter?

**Review by bbrks — COMMENTED** [review]
Left some questions inline

File: rest/handler.go (hunk 3 above)
Comment (torcolvin) [reply rest/handler.go:603]: These are actually not done with `http.ResponseWriter` so potentially I don't even have to skip this here, I'm not sure what is more clear.

File: base/stats.go (hunk 12 above)
Comment (torcolvin) [reply base/stats.go]: Split to https://issues.couchbase.com/browse/CBG-3163

Comment (torcolvin) [reply base/stats.go]: renamed to PublicRestBytesWritten to match NumPublicRestRequests

**Review by adamcfraser — APPROVED** [review]
